src/twisted/protocols/test/test_tls.py

```diff
@@ -1899,108 +1922,111 @@ def dataReceived(self, data):
         wrapperFactory = TLSMemoryBIOFactory(clientContextFactory, True, clientFactory)
         sslClientProtocol = wrapperFactory.buildProtocol(None)
 
         serverDataReceived = Deferred()
         serverFactory = ServerNegotiationFactory(serverProtocols)
-        serverFactory.protocol = lambda: NotifyingSender(serverDataReceived)
+        serverFactory.protocol = lambda: NotifyingSender(
+            serverDataReceived, b"data from server"
+        )
 
         serverContextFactory = ServerTLSContext()
         wrapperFactory = TLSMemoryBIOFactory(serverContextFactory, False, serverFactory)
         sslServerProtocol = wrapperFactory.buildProtocol(None)
 
         loopbackAsync(sslServerProtocol, sslClientProtocol)
-        return (
-            sslClientProtocol,
-            sslServerProtocol,
-            clientDataReceived,
-            serverDataReceived,
+        deferred = gatherResults(
+            [clientDataReceived, serverDataReceived], consumeErrors=True
         )
+        deferred.addCallback(lambda _: (sslClientProtocol, sslServerProtocol))
+        deferred.addErrback(lambda firstError: firstError.value.args[0])
+        return deferred
 
     def test_negotiationWithNoProtocols(self):
         """
         When factories support L{IProtocolNegotiationFactory} but don't
         advertise support for any protocols, no protocols are negotiated.
         """
-        (
-            client,
-            server,
-            clientDataReceived,
-            serverDataReceived,
-        ) = self.handshakeProtocols([], [])
+        deferred = self.handshakeProtocols([], [])
 
-        def checkNegotiatedProtocol(ignored):
+        def checkNegotiatedProtocol(result):
+            client, server = result
             self.assertEqual(client.negotiatedProtocol, None)
             self.assertEqual(server.negotiatedProtocol, None)
 
-        clientDataReceived.addCallback(lambda ignored: serverDataReceived)
-        serverDataReceived.addCallback(checkNegotiatedProtocol)
-
-        return clientDataReceived
+        return deferred.addCallback(checkNegotiatedProtocol)
 
     def test_negotiationWithProtocolOverlap(self):
         """
         When factories support L{IProtocolNegotiationFactory} and support
         overlapping protocols, the first protocol is negotiated.
         """
-        (
-            client,
-            server,
-            clientDataReceived,
-            serverDataReceived,
-        ) = self.handshakeProtocols([b"h2", b"http/1.1"], [b"h2", b"http/1.1"])
+        deferred = self.handshakeProtocols(
+            clientProtocols=[b"http/1.1", b"h2", b"client"],
+            serverProtocols=[b"h2", b"http/1.1", b"server"],
+        )
 
-        def checkNegotiatedProtocol(ignored):
+        def checkNegotiatedProtocol(result):
+            client, server = result
             self.assertEqual(client.negotiatedProtocol, b"h2")
             self.assertEqual(server.negotiatedProtocol, b"h2")
```

Comment:
So this seems to be the test where the explicit `_setAcceptableProtocols` call matters.  It seems like our ALPN implementation is doing what we think it should be doing - ie, even without the explicit `_setAcceptableProtocols` in the tests, the implementation makes sure to call the right OpenSSL ALPN APIs.  But then they don't seem to have any effect.

I'll try to dig into this more, I guess.  It seems like we've misunderstood the OpenSSL ALPN APIs or they have a bug or they've subtly changed their behavior somehow.